File: examples/scripts/ts/fetch_subject.ts
/**
 * Model Health TypeScript SDK — fetch a subject by ID.
 * Mirrors examples/python/fetch_subject.py.
 *
 * Usage:
 *   npx tsx fetch_subject.ts [<api_key>]
 */

import { ModelHealthClient } from '@modelhealth/modelhealth';
import type { Subject } from '@modelhealth/modelhealth';
import { loadApiKey, pickOne, closePrompts } from './_shared.js';

async function connect(apiKey: string): Promise<ModelHealthClient> {
  console.log('Connecting...');
  const client = new ModelHealthClient({ apiKey, autoInit: false });
  await client.init();
  return client;
}

async function pickSubject(client: ModelHealthClient): Promise<Subject> {
  console.log('\nFetching subjects...');
  const subjects = await client.subjectList();

  if (!subjects.length) {
    console.error('No subjects found.');
    process.exit(1);
  }

  console.log();
  const subject = await pickOne(subjects, 'Select subject', s => `${s.name}  (ID ${s.id})`);
  console.log(`  Selected: ${subject.name}`);
  return subject;
}

function printSubject(subject: Subject): void {
  console.log(`  Name:             ${subject.name}`);
  console.log(`  Weight:           ${subject.weight ?? '(none)'}`);
  console.log(`  Height:           ${subject.height ?? '(none)'}`);
  console.log(`  Birth year:       ${subject.birthYear ?? '(none)'}`);
  console.log(`  Age:              ${subject.age ?? '(none)'}`);
  console.log(`  Gender:           ${subject.gender}`);
  console.log(`  Sex at birth:     ${subject.sexAtBirth}`);
  console.log(`  Characteristics:  ${subject.characteristics || '(none)'}`);
}

async function main() {
  const args = process.argv.slice(2);
  const client = await connect(loadApiKey(args[0]));
  const subject = await pickSubject(client);

  console.log(`\nFetching subject ${subject.id}...`);
  const fetched = await client.fetchSubject(subject.id);
  console.log();
  printSubject(fetched);
}

main().catch(err => { console.error(`Error: ${err.message ?? err}`); process.exit(1); })
  .finally(closePrompts);

File: examples/scripts/python/fetch_subject.py
#!/usr/bin/env python3
"""Model Health Python SDK — fetch a subject by ID demo.

Walks through resolving a subject ID into full subject details:
  1. Select a subject (to obtain a valid ID)
  2. Fetch that subject by ID via `fetch_subject` and print its details

Usage:
    fetch_subject.py [<api_key>]
"""

import sys

from docopt import docopt

from modelhealth import ModelHealthError, ModelHealthClient
from _prompts import pick_one
from _utils import load_api_key


# ---------------------------------------------------------------------------
# Setup
# ---------------------------------------------------------------------------

def _connect(api_key):
    print("Connecting...")
    try:
        return ModelHealthClient(api_key)
    except ModelHealthError as exc:
        sys.exit(f"Failed to initialise: {exc}")


# ---------------------------------------------------------------------------
# Subject selection / lookup
# ---------------------------------------------------------------------------

def _pick_subject(client):
    print("\nFetching subjects...")
    try:
        subjects = client.subject_list()
    except ModelHealthError as exc:
        sys.exit(f"Failed to fetch subjects: {exc}")

    if not subjects:
        sys.exit("No subjects found.")

    print()
    subject = pick_one(subjects, "Select subject", lambda s: f"{s.name}  (ID {s.id})")
    print(f"  Selected: {subject.name}")
    return subject


def _fetch_subject(client, subject_id):
    print(f"\nFetching subject {subject_id}...")
    try:
        return client.fetch_subject(subject_id)
    except ModelHealthError as exc:
        sys.exit(f"Failed to fetch subject: {exc}")


def _print_subject(subject):
    print(f"  Name:             {subject.name}")
    print(f"  Weight:           {subject.weight if subject.weight is not None else '(none)'}")
    print(f"  Height:           {subject.height if subject.height is not None else '(none)'}")
    print(f"  Birth year:       {subject.birth_year if subject.birth_year is not None else '(none)'}")
    print(f"  Age:              {subject.age if subject.age is not None else '(none)'}")
    print(f"  Gender:           {subject.gender}")
    print(f"  Sex at birth:     {subject.sex_at_birth}")
    print(f"  Characteristics:  {subject.characteristics or '(none)'}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main(api_key):
    client = _connect(api_key)
    subject = _pick_subject(client)

    fetched = _fetch_subject(client, subject.id)
    print()
    _print_subject(fetched)


if __name__ == "__main__":
    args = docopt(__doc__)
    main(load_api_key(args["<api_key>"]))
